### ISSUE TITLE
No notifications on Rocket.chat+ Android

### ISSUE THREAD

**Issue by mhham** [RocketChat/Rocket.Chat.Android#254]
Your Rocket.Chat.Android version: 1.0.8

No notifications at all on several devices (Lg G4, Honor 5C, Oneplus 1)

**Comment by reffu42** [comment]
I have the same issue (Galaxy S6 on Nougat) as do several other people at my organization.

I'm on 1.0.8 and have already tried clearing app data and reinstalling, neither of which have fixed it

**Comment by laggedHero** [comment]
@mhham @reffu42 Did you guys tried the test push (on the administration menu)?

Could you to that and tell me the results?

Are you guys using our gateway?

**Comment by mhham** [comment]
I am using the Rocketchat+ client (native Android, not the webview wrapper), with the demo.rocket.chat server.

**Comment by mhham** [comment]
Still no news on the issue

**Comment by cyclops24** [comment]
I have same issue and notification not working for me on an Android 7.0.0 mobile phone. (LG K10 2017)

**Comment by geekgonecrazy** [comment]
Please goto your push settings on your server and ensure you are using our push gateway as requested by @laggedHero above :top: 

This is important for debugging.

Also can close the app on your phone and desktop then login with another user and see if it shows your users status as green.  And let us know :+1:

**Comment by rafaelks** [comment]
@cyclops24 @mhham Is this issue still happening?

**Comment by sttyru** [comment]
@rafaelks 
Hello! I can to confirm that this problem still happening...
I'm use https://gateway.rocket.chat
At the server log found next message:
Push: send message "@MYUSERNAME" via query { userId: 'eHhuF1m99baHRDjxBg' }

**Comment by rafaelks** [comment]
@sttyru Does it have some error information? Does it work on other devices? What's your server version?

**Comment by congnt24** [comment]
After debugging the code, I found that if you use default gateway: https://gateway.rocket.chat when getting public-settings from your server, Push_gcm_project_number(GCM SenderId) will return null. So you can't receive notification from your device.
I have some solution for this:
1. You can use your own gateway by creating on firebase and pass your "GCM API Key" and "GCM Project Number" in settings/push.
2. Set default GCM Sender Id in code. It's will be used if the Push_gcm_project_number from the server is null.

**Comment by geekgonecrazy** [comment]
@congnt24 are you getting push notifications with our rocketchat+ app in the store or are you compiling the app your self?

If you are compiling your self then yes you absolutely have to use your own credentials.

But if you are using ours to my knowledge you will have to use our gateway.

**Comment by congnt24** [comment]
@geekgonecrazy 
I using rocketchat+ app in the store, it works with both, my own credentials and default gateway.
then I am compiling for myself and only receive notification if I use default gateway.
So, I think other people build their own app but they use default gateway, the Sender ID will null and they can't receive notification.

**Comment by rafaelks** [comment]
@congnt24 If you're compiling the app yourself, you'll need to use your own credentials on the server. If you're using the App Store version of the app, you can use our gateway to send the notifications.

**Comment by xenithorb** [comment]
It's not that you CAN use the R.C. gateway when the app isn't self-compiled, you **must**. I think that gets lost on people. The public app is simply built with the gateway's credentials, just like the cordova version.

I tested the Play store version against my own server recently and push works, btw.

**Comment by geekgonecrazy** [comment]
Yes exactly!  If you use our published apps you **must** use the gateway for notifications.  If you compile your own.. you **cannot** use the gateway for notifications.

**Comment by antgel** [comment]
@geekgonecrazy I'm plotting a doc patch on this subject. Is the following true?

* If you use official RC app, you must use the official RC gateway.
* If you fork, compile, and release your own RC app, you must do one of the following:
  * Set up your own gateway (if your app may connect to different RC servers).
  * Not use a gateway at all (if your app is only for internal use and only connects to a specific RC instance). In this case you need to install certificates on your RC instance.

**Comment by geekgonecrazy** [comment]
That appears to be correct!

**Comment by xenithorb** [comment]
@antgel Also important to mention that at the point where they desire to compile their own apps and distribute them, that would be when they start getting their own API keys and such. (As a process of modifying the app to distribute, and also adding those credentials to the server)

**Comment by antgel** [comment]
@xenithorb Just when I thought I'd grokked it... So, you only need your own API key (I'm talking GCM, I don't know about APN) when you compile your own app? Does this mean that the `GCM API Key` and `GCM Project Number` fields are meaningless when using the official RC app?

**Comment by xenithorb** [comment]
@antgel That is correct

**Comment by geekgonecrazy** [comment]
Ah sorry.  Yes instead of your two sub steps it would be get gcm tokens for push notifications

**Comment by antgel** [comment]
@geekgonecrazy What? :)

**Comment by geekgonecrazy** [comment]
* If you use official RC app, you must use the official RC gateway.
* If you fork, compile, and release your own RC app, you must do one of the following:
  * Get GCM tokens from Google and compile them into the app
  * Set them in the push section of Rocket.Chat server admin panel

**Comment by antgel** [comment]
@geekgonecrazy I'm happy we're having a conversation about this topic, which seems to confuse people. We're getting there. :)

Are you sure it should still read do *one* of the following? Surely it should be:
* Get GCM tokens from Google and compile them into the app
AND
* Set them in the push section of Rocket.Chat server admin panel?

Also, when would there be a use case to use a gateway that *isn't* https://gateway.rocket.chat, as opposed performing the two steps above?

**Comment by geekgonecrazy** [comment]
So if you intend to publish an app specific for your server and your server only you have to go through the head ache that we have to.

If you aren't publishing your own app and intend for people to be able to use our app and have multiple servers.  You would use our gateway and not bother with apps, and also wouldn't touch push settings at all.

Really for people building their own app, there is no point in a gateway, because it would only work with servers pointing to your gateway and with your app.  If someone tried to add another rocket.chat server they would not be able to get push notifications for that server.

I don't think anyone wants the headache of putting that level of confusion on your users.  Major support headache

**Comment by antgel** [comment]
So, can we conclude that the only sane gateway in RC-world is indeed https://gateway.rocket.chat, and there is no meaningful use case for another gateway?

**Comment by nodens** [comment]
Hi, just tested this as well. RC+ app from play store (1.0.15), server 0.56.0.

We use the default gateway https://gateway.rocket.chat.

If I test the push with the button in administration -> push, it works, but I never receive any push when people send me a direct message or highlight me in a channel. Also, no entry in the log even with debug enabled in the push settings.

I managed to get it working once, when I made sure I was logged off everywhere but on the app, but even that didn't work after a while (also making sure web, desktop etc is closed wouldn't be very user-friendly, with the electron app (beta) this part works).

That's too bad, it works so much better and faster than the electron version otherwise...

**Comment by nodens** [comment]
More info, after leaving it for a while: it sometimes work, sometimes doesn't... But I couldn't reproduce the working or non-working reliably... :-/

**Comment by geekgonecrazy** [comment]
@nodens the whole push notification thing is very dependant on your connection status elsewhere.  So if you were just using the app on the desktop its not going to alert you on your phone, its going to alert your desktop.

We have some work in progress server side to improve that

**Comment by jasonfah** [comment]
Maybe this is helpful? https://twitter.com/mathowie/status/837735473745289218

**Comment by Matt-9999** [comment]
@geekgonecrazy  Since the notification icon disappears, once I set the app in background, is there a plan to work on the app-side also?

**Comment by geekgonecrazy** [comment]
i'm not sure I understand the question..  Can you rephrase?

**Comment by Matt-9999** [comment]
ok, I start the Rocket.Chat app on Android, go to "My Account/Preferences" menu, press on "TEST DESKTOP NOTIFICATION".
The notification appears in the status bar as the most left icon on my phone.
Then I push the home button to put the Rocket.Chat app to the background.
The notification icon disappears, but it should stay in the status bar along with the notification icons from other apps.

**Comment by rafaelks** [comment]
Guys, thanks for all your feedback! We're investigating the issue and we found a problem in the Android app that let's the socket connection opened, telling the server to not send push notifications to that device.

We're working to fix this issue ASAP, thanks for your patience!

**Comment by rafaelks** [comment]
This one should be fixed already! Thank you all for your patience! 👍

**Comment by phlegx** [comment]
@rafaelks Does not seem to work for me on Android with version RC version `0.59.3`. There are no notification icons appearing on the phone when somebody writes me, neither from a normal user nor from the live chat. 

Test desktop notification works as well for me, but I experience the same behavior as @Matt-9999

**Comment by Alseenrodelap** [comment]
I still have issues too. My info:
RC version: 59.3, running in docker.
Ubuntu 16.04 lts.
Android app installed via Playstore.
Using the https://gateway.rocket.chat gateway.

When I press the 'test push' button, I get a push message on my phone like it should. But when I log out, and log in with another user and mention myself or send myself a DM, I don't get any push messages.

**Comment by rafaelks** [comment]
@Alseenrodelap Is your other user offline when you try to mention him?

**Comment by Alseenrodelap** [comment]
Before I logged out I set myself to 'not available'. Then I manually log out. Then I log in with another user and I see that my first user is still set on 'not available' (yellow dot). So I think, maybe the app is keeping me logged in. So I open my Android app, log out, close the app and try again. Still my first user has a yellow dot, even after hard rebooting the server. I am sure that there are no other sessions open.

Also, I think when I put myself on 'not available', it should already start sending push messages to my phone. I also selected 'all messages' to invoke a push message, but that doesn't help either.

**Comment by Alseenrodelap** [comment]
Even after a night of sleep the primary user has a yellow dot and doesn't receive push messages.

**Comment by abkil** [comment]
Hello, same issue here.
OS : Debian 9 (fresh install)
Version : 59.3 installed with the snap package
Official android app + default gateway

**Comment by deenski** [comment]
same here, deb 8, snap, default app and gateway

**Comment by horstepipe** [comment]
yeah, my android driven users complain about it, too. Ubuntu, snap, official app + default gateway.
IOS works fine btw

**Comment by rafaelks** [comment]
We're working in a fix for notifications, let's see if it works for you guys after that.

**Comment by horstepipe** [comment]
problem persists on server v0.60.2.
I compared the push tokens in the rc.server log and with this app on the client device, they are different:
https://play.google.com/store/apps/details?id=com.twobasetechnologies.devicetokenfinder

Tried to logout the user from all devices and reinstalled rc+ on the android device. The old (different) token was removed (as I saw in rc log), but messages are now being sent to a new different one

**Comment by rafaelks** [comment]
@horstepipe I believe these tokens are suppose to be different. Push token is something unique per application, this token the app is showing you is the unique identifier of the device, something used to identify the device, not to send notifications. Right?

**Comment by horstepipe** [comment]
Ok thanks, hmm then I have no idea what’s going wrong there.

**Comment by atc07** [comment]
@rafaelks this still seems like an issue. Why was this closed then? As it seems like push notifications for Rocket Chat Android doesn't work unless we edit the code to work with Firebase or something else. Can you confirm either way?

**Comment by rafaelks** [comment]
@atc07 It depends... are you forking the repo? And are you using our gateway?

**Comment by atc07** [comment]
Yes we are forking the repo and using your gateway.

**Comment by rafaelks** [comment]
@atc07 Did you change the package name? In this case, I believe you'll have to use your own push settings.

**Comment by AnshD12** [comment]
hello everyone I tried all the steps mentioned in rocket chat docs for push notification but its still not working its showing and error `there are no tokens for this user` let me tell you I am trying this for personal android project based on rocket chat apis
so can anyone tell me
@rafaelks if you can help it would be great 
1. how this token gets generated is there a method or I need to that manually
2. do I need to add google-services.json to android project
3. anything else I need to do apart from given steps on docs for push notification for android

**Comment by atc07** [comment]
@rafaelks yes already changed packaged name and we are using our own push settings.

**Comment by rafaelks** [comment]
@AnshD12 Already replied to your in our chat channel.

@atc07 Did you update the files in the app regarding Google services?

**Comment by magikmw** [comment]
I'm encountering this issue.
Server: 0.62.2, App RocketChat+ v1.1.0 on Android 8.0
iOS app on the same server works fine with push, the Android app doesn't receive notifications.
Default settings for push notifications, test says 'sending to 2 devices'.

**Comment by icemansparks** [comment]
Similar Issue: seemingly no mobile notifications (regardless of settings).
used: MacBook and iPhone (rocket chat +) self hosted.

reproduced: 
- no mobile notifications while desktop client is active - notifications for desktop work fine in this case.
- all mobile notifications work find when desktop app is completely closed (not just hidden) - even when the app is not running in background on the phone.

How it works now, the functionality does not make any sense to me. I will be using the app on both systems a lot and don't want to remember closing it every time when I will not use the macbook for a few minutes while I'm out of the office...

**Comment by eeldwin** [comment]
I am experiencing similar problem with icemansparks

**Comment by sbernhard** [comment]
Would be great to have a solution for Android soon. Doesn't make much fun without push notifications.

**Comment by rafaelks** [comment]
This issue is related to version 1.1.0 guys. Are you all still having issues on version 2.0.0? If yes, what's the issue?

**Comment by reetp** [comment]
Think so https://github.com/RocketChat/Rocket.Chat.Android/issues/1111

**Comment by atc07** [comment]
We had the same issues and believe we fixed with this commit: https://github.com/RocketChat/Rocket.Chat/pull/9887

Take a look at and pull whatever you need. We are able to get push notifications working on our end with these changes on Android.

**Comment by reetp** [comment]
Thanks.
I use docker for Rocket itself and have no build system for Android so I am stuck waiting for the devs to fix this.

**Comment by iliaik** [comment]
The same problem with Android.
Also on Ubuntu with app..

**Comment by reetp** [comment]
Yup. Half finished code and bugs on this all over the place (can't see #9987 has ever been included) and still no notifications.

Makes the Android app, and Rocket, effectively useless. Have been here for a good few months.

I can only presume the devs paying markets don't require or depend on this sort of functionality, or have a "different direction of travel", as it seems to have pretty low priority.

**Comment by rafaelks** [comment]
@reetp @ilyaikik This is a not a low priority for us. What issue are you having? This topic is about 1.1.0.

**Comment by rafaelks** [comment]
I'm locking this issue here, since it's related to 1.1.0. If you're having issues with 2.0.0+ that aren't open yet, please create a new issue. 👍
